src/wagtail_2fa/middleware.py

```diff
@@ -92,3 +92,29 @@ def _get_paths(self, route_names):
             except NoReverseMatch:
                 pass
         return results
+
+
+class VerifyUserPermissionsMiddleware(VerifyUserMiddleware):
+    """A variant of VerifyUserMiddleware which makes 2FA optional."""
+
+    def process_request(self, request):
+        result = super().process_request(request)
+
+        # Add an attribute to the user so we can easily determine if 2FA should
+        # be enabled for them.
+        if request.user.has_perms(["wagtailadmin.enable_2fa"]):
+            request.user.enable_2fa = True
+        else:
+            request.user.enable_2fa = False
+
+        return result
+
+    def _require_verified_user(self, request):
+        result = super()._require_verified_user(request)
+        user = request.user
+
+        # Don't require verification if the user has 2FA disabled
+        if not user.has_perms(["wagtailadmin.enable_2fa"]):
```

Comment:
think we can just use ``request.user`` here , then you don't have to lookup ``user`` first